id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/controller/IdRepoDraftController.java

```diff
@@ -0,0 +1,171 @@
+					request.getRequest().getRegistrationId(), IdType.RID, "Update draft requested");
+		}
+	}
+
+	@PreAuthorize("hasAnyRole('REGISTRATION_PROCESSOR')")
+	@GetMapping(path = "/publishDraft/{registrationId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<IdResponseDTO> publishDraft(@PathVariable String registrationId) throws IdRepoAppException {
+		try {
+			validateRegistrationId(registrationId);
+			return new ResponseEntity<>(draftService.publishDraft(registrationId), HttpStatus.OK);
+		} catch (IdRepoAppException e) {
+			auditHelper.auditError(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.PUBLISH_DRAFT_REQUEST_RESPONSE, registrationId,
+					IdType.RID, e);
+			mosipLogger.error(IdRepoSecurityManager.getUser(), ID_REPO_DRAFT_CONTROLLER, "publishDraft", e.getMessage());
+			throw new IdRepoAppException(e.getErrorCode(), e.getErrorText(), e);
+		} finally {
+			auditHelper.audit(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.CREATE_DRAFT_REQUEST_RESPONSE, registrationId,
+					IdType.RID, "Publish draft requested");
+		}
+	}
+
+	@PreAuthorize("hasAnyRole('REGISTRATION_PROCESSOR')")
+	@DeleteMapping(path = "/discardDraft/{registrationId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<IdResponseDTO> discardDraft(@PathVariable String registrationId) throws IdRepoAppException {
+		try {
+			validateRegistrationId(registrationId);
+			return new ResponseEntity<>(draftService.discardDraft(registrationId), HttpStatus.OK);
+		} catch (IdRepoAppException e) {
+			auditHelper.auditError(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.DISCARD_DRAFT_REQUEST_RESPONSE, registrationId,
+					IdType.RID, e);
+			mosipLogger.error(IdRepoSecurityManager.getUser(), ID_REPO_DRAFT_CONTROLLER, "discardDraft", e.getMessage());
+			throw new IdRepoAppException(e.getErrorCode(), e.getErrorText(), e);
+		} finally {
+			auditHelper.audit(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.DISCARD_DRAFT_REQUEST_RESPONSE, registrationId,
+					IdType.RID, "Discard draft requested");
+		}
+	}
+
+	@PreAuthorize("hasAnyRole('REGISTRATION_PROCESSOR')")
+	@GetMapping(path = "/hasDraft/{registrationId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<IdResponseDTO> hasDraft(@PathVariable String registrationId) throws IdRepoAppException {
+		try {
+			validateRegistrationId(registrationId);
+			return new ResponseEntity<>(draftService.hasDraft(registrationId), HttpStatus.OK);
+		} catch (IdRepoAppException e) {
+			auditHelper.auditError(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.HAS_DRAFT_REQUEST_RESPONSE, registrationId,
+					IdType.RID, e);
+			mosipLogger.error(IdRepoSecurityManager.getUser(), ID_REPO_DRAFT_CONTROLLER, "hasDraft", e.getMessage());
+			throw new IdRepoAppException(e.getErrorCode(), e.getErrorText(), e);
+		} finally {
+			auditHelper.audit(AuditModules.ID_REPO_CORE_SERVICE, AuditEvents.HAS_DRAFT_REQUEST_RESPONSE, registrationId,
```

Comment:
can we have the audit more in the service layer rather than here?

Reply:
Having audit in controller helps separate business logic in service layer and auditing logic. And also it makes easier to update auditing logs from here. The same has been followed in IDRepo and IDA.